Add support for Optimized Hash under MSVC

fsfod opened the pull request:
Add support for Optimized Hash under Visual c++ on x64 for both cl and clang-cl. I also changed lj_str_hashfn size parameter to MSize since string can't be larger than that and its just be implicitly truncated in lj_str_hash.c which was causing lots of warnings.

fsfod committed:
Add support for building LJ_HAS_OPTIMISED_HASH under MSVC

Signed-off-by: Thomas Fransham <fsfod11@gmail.com>

fsfod committed:
Fix implicit 64 to 32 bit truncation warnings in lj_str_hash.c

Changed lj_str_hashfn's size parameter to MSize instead size_t since it just
being truncated in lj_str_hash.c

Signed-off-by: Thomas Fransham <fsfod11@gmail.com>

siddhesh commented:
Thanks, does the MSVC build process use Makefiles at all?  If not, then you'll need to patch msvcbuild.bat to build lj_str_hash.c with -msse4.2 (rather the msvc equivalent) and then rest with default x86_64 flags so that the binary is portable across x86_64 machines.

fsfod committed:
Don't rely on the implicit MSVC compiler declaration of _mm_crc32_u32\_mm_crc32_u64.

Use the correct intrinsic header for SSE 4.2 nmmintrin.h instead of smmintrin.h
which is for SSE 4.1.
Fix new warnings from CRC intrinsics having a proper declaration.

Signed-off-by: Thomas Fransham <fsfod11@gmail.com>

fsfod committed:
Set LJ_HAS_OPTIMISED_HASH when compiled under x64 MSVC.

Just remove the x86 check i added in lj_str_hash.c it is already checked
by lj_arch.h.

Signed-off-by: Thomas Fransham <fsfod11@gmail.com>

fsfod commented:
MSVC build process only uses msvcbuild.bat and just uses file name globbing to [pick the source files to build](https://github.com/moonjit/moonjit/blob/master/src/msvcbuild.bat#L83), so it was already building lj_str_hash.c .

MSVC only has the [/arch switch](https://docs.microsoft.com/en-us/cpp/build/reference/arch-x64?view=vs-2019) and the only options for it are AVX\AVX2\AVX512 when building for x64. I did make a comment on this in the code that MSVC allows you to use intrinsics from any of the SSE versions without the /arch switch being set to a AVX or higher.

siddhesh commented:
Merged.  I trust your understanding of MSVC because it's definitely better than mine :)
